# Finetune with LoRA

In [1]:
# pip install transformers datasets lightning watermark

In [2]:
%load_ext watermark
%watermark --conda -p torch,transformers,datasets,lightning

torch       : 2.1.1+cu121
transformers: 4.36.2
datasets    : 2.16.1
lightning   : 2.1.2

conda environment: n/a



# 1 Loading the dataset into DataFrames

In [2]:
import os
from datasets import load_dataset

import lightning as L
from lightning.pytorch.loggers import CSVLogger
from lightning.pytorch.callbacks import ModelCheckpoint

import pandas as pd
import torch

from local_dataset_utilities import download_dataset, load_dataset_into_to_dataframe, partition_dataset
from local_dataset_utilities import IMDBDataset

In [3]:
if not torch.cuda.is_available():
    print("Please switch to a GPU machine before running this notebook.")


In [4]:
files = ("test.csv", "train.csv", "val.csv")
download = True

for f in files:
    if not os.path.exists(os.path.join("data", f)):
        download = False

if download is False:
    download_dataset()
    df = load_dataset_into_to_dataframe()
    partition_dataset(df)

In [5]:
df_train = pd.read_csv(os.path.join("data", "train.csv"))
df_val = pd.read_csv(os.path.join("data", "val.csv"))
df_test = pd.read_csv(os.path.join("data", "test.csv"))

# 2 Tokenization and Numericalization

**Load the dataset via `load_dataset`**

In [6]:
imdb_dataset = load_dataset(
    "csv",
    data_files={
        "train": os.path.join("data", "train.csv"),
        "validation": os.path.join("data", "val.csv"),
        "test": os.path.join("data", "test.csv"),
    },
)

print(imdb_dataset)

Extracting data files:   0%|          | 0/3 [00:00<?, ?it/s]

Generating train split: 0 examples [00:00, ? examples/s]

Generating validation split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['index', 'text', 'label'],
        num_rows: 35000
    })
    validation: Dataset({
        features: ['index', 'text', 'label'],
        num_rows: 5000
    })
    test: Dataset({
        features: ['index', 'text', 'label'],
        num_rows: 10000
    })
})


**Tokenize the dataset**

In [7]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")
print("Tokenizer input max length:", tokenizer.model_max_length)
print("Tokenizer vocabulary size:", tokenizer.vocab_size)

D:\anaconda\lib\site-packages\huggingface_hub\file_download.py:133: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\57264\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Tokenizer input max length: 512
Tokenizer vocabulary size: 30522


In [8]:
def tokenize_text(batch):
    return tokenizer(batch["text"], truncation=True, padding=True)

In [9]:
imdb_tokenized = imdb_dataset.map(tokenize_text, batched=True, batch_size=None)

Map:   0%|          | 0/35000 [00:00<?, ? examples/s]

Map:   0%|          | 0/5000 [00:00<?, ? examples/s]

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [10]:
del imdb_dataset

In [11]:
imdb_tokenized.set_format("torch", columns=["input_ids", "attention_mask", "label"])

In [12]:
os.environ["TOKENIZERS_PARALLELISM"] = "false"

# 3 Set Up DataLoaders

In [13]:
from torch.utils.data import DataLoader, Dataset


class IMDBDataset(Dataset):
    def __init__(self, dataset_dict, partition_key="train"):
        self.partition = dataset_dict[partition_key]

    def __getitem__(self, index):
        return self.partition[index]

    def __len__(self):
        return self.partition.num_rows

In [30]:
train_dataset = IMDBDataset(imdb_tokenized, partition_key="train")
val_dataset = IMDBDataset(imdb_tokenized, partition_key="validation")
test_dataset = IMDBDataset(imdb_tokenized, partition_key="test")

train_loader = DataLoader(
    dataset=train_dataset,
    batch_size=12,
    shuffle=True, 
    num_workers=0
)

val_loader = DataLoader(
    dataset=val_dataset,
    batch_size=12,
    num_workers=0
)

test_loader = DataLoader(
    dataset=test_dataset,
    batch_size=12,
    num_workers=0
)

# 4 Initializing DistilBERT

In [45]:
from transformers import AutoModelForSequenceClassification


model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased", num_labels=2)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_projector.bias', 'vocab_layer_norm.bias', 'vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_transform.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'pre_classifier.bias', 'classifier.bias', 'classifier.we

**Freeze all layers**

In [46]:
for param in model.parameters():
    param.requires_grad = False

**Add LoRA layers**

In [47]:
model

DistilBertForSequenceClassification(
  (distilbert): DistilBertModel(
    (embeddings): Embeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (transformer): Transformer(
      (layer): ModuleList(
        (0-5): 6 x TransformerBlock(
          (attention): MultiHeadSelfAttention(
            (dropout): Dropout(p=0.1, inplace=False)
            (q_lin): Linear(in_features=768, out_features=768, bias=True)
            (k_lin): Linear(in_features=768, out_features=768, bias=True)
            (v_lin): Linear(in_features=768, out_features=768, bias=True)
            (out_lin): Linear(in_features=768, out_features=768, bias=True)
          )
          (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (ffn): FFN(
            (dropout): Dropout(p=0.1, inplace=False)
 

In [34]:
class LoRALayer(torch.nn.Module):
    def __init__(self, in_dim, out_dim, rank, alpha):
        super().__init__()
        std_dev = 1 / torch.sqrt(torch.tensor(rank).float())
        self.W_a = torch.nn.Parameter(torch.randn(in_dim, rank) * std_dev)
        self.W_b = torch.nn.Parameter(torch.zeros(rank, out_dim))
        self.alpha = alpha

    def forward(self, x):
        x = self.alpha * (x @ self.W_a @ self.W_b)
        return x


class LinearWithLoRA(torch.nn.Module):
    def __init__(self, linear, rank, alpha):
        super().__init__()
        self.linear = linear
        self.lora = LoRALayer(
            linear.in_features, linear.out_features, rank, alpha
        )

    def forward(self, x):
        return self.linear(x) + self.lora(x)

In [35]:
from functools import partial

lora_r = 8
lora_alpha = 16
lora_dropout = 0.05
lora_query = True
lora_key = False
lora_value = True
lora_projection = False
lora_mlp = False
lora_head = False

layers = []

assign_lora = partial(LinearWithLoRA, rank=lora_r, alpha=lora_alpha)

for layer in model.distilbert.transformer.layer:
    if lora_query:
        layer.attention.q_lin = assign_lora(layer.attention.q_lin)
    if lora_key:
        layer.attention.k_lin = assign_lora(layer.attention.k_lin)
    if lora_value:
        layer.attention.v_lin = assign_lora(layer.attention.v_lin)
    if lora_projection:
        layer.attention.out_lin = assign_lora(layer.attention.out_lin)
    if lora_mlp:
        layer.ffn.lin1 = assign_lora(layer.ffn.lin1)
        layer.ffn.lin2 = assign_lora(layer.ffn.lin2)
if lora_head:
    model.pre_classifier = assign_lora(model.pre_classifier)
    model.classifier = assign_lora(model.classifier)

In [48]:
model

DistilBertForSequenceClassification(
  (distilbert): DistilBertModel(
    (embeddings): Embeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (transformer): Transformer(
      (layer): ModuleList(
        (0-5): 6 x TransformerBlock(
          (attention): MultiHeadSelfAttention(
            (dropout): Dropout(p=0.1, inplace=False)
            (q_lin): Linear(in_features=768, out_features=768, bias=True)
            (k_lin): Linear(in_features=768, out_features=768, bias=True)
            (v_lin): Linear(in_features=768, out_features=768, bias=True)
            (out_lin): Linear(in_features=768, out_features=768, bias=True)
          )
          (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (ffn): FFN(
            (dropout): Dropout(p=0.1, inplace=False)
 

In [49]:
# Check if linear layers are frozen
for name, param in model.named_parameters():
    print(f"{name}: {param.requires_grad}")

distilbert.embeddings.word_embeddings.weight: False
distilbert.embeddings.position_embeddings.weight: False
distilbert.embeddings.LayerNorm.weight: False
distilbert.embeddings.LayerNorm.bias: False
distilbert.transformer.layer.0.attention.q_lin.weight: False
distilbert.transformer.layer.0.attention.q_lin.bias: False
distilbert.transformer.layer.0.attention.k_lin.weight: False
distilbert.transformer.layer.0.attention.k_lin.bias: False
distilbert.transformer.layer.0.attention.v_lin.weight: False
distilbert.transformer.layer.0.attention.v_lin.bias: False
distilbert.transformer.layer.0.attention.out_lin.weight: False
distilbert.transformer.layer.0.attention.out_lin.bias: False
distilbert.transformer.layer.0.sa_layer_norm.weight: False
distilbert.transformer.layer.0.sa_layer_norm.bias: False
distilbert.transformer.layer.0.ffn.lin1.weight: False
distilbert.transformer.layer.0.ffn.lin1.bias: False
distilbert.transformer.layer.0.ffn.lin2.weight: False
distilbert.transformer.layer.0.ffn.lin2.bi

In [50]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)


print("Total number of trainable parameters:", count_parameters(model))

Total number of trainable parameters: 0


# 5 Finetuning

**Wrap in LightningModule for Training**

In [54]:
from local_model_utilities import CustomLightningModule

lightning_model = CustomLightningModule(model)

In [55]:
callbacks = [
    ModelCheckpoint(
        save_top_k=1, mode="max", monitor="val_acc"
    )  # save top 1 model
]
logger = CSVLogger(save_dir="logs/", name="my-model")

In [56]:
trainer = L.Trainer(
    max_epochs=3,
    callbacks=callbacks,
    accelerator="gpu",
    precision="16-mixed",
    devices=1,
    logger=logger,
    log_every_n_steps=10,
)

Using 16bit Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs


In [57]:
import time
start = time.time()

trainer.fit(model=lightning_model,
            train_dataloaders=train_loader,
            val_dataloaders=val_loader)

end = time.time()
elapsed = end - start
print(f"Time elapsed {elapsed/60:.2f} min")

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name     | Type                                | Params | Mode 
-------------------------------------------------------------------------
0 | model    | DistilBertForSequenceClassification | 67.0 M | eval 
1 | val_acc  | MulticlassAccuracy                  | 0      | train
2 | test_acc | MulticlassAccuracy                  | 0      | train
-------------------------------------------------------------------------
0         Trainable params
67.0 M    Non-trainable params
67.0 M    Total params
267.820   Total estimated model params size (MB)
2         Modules in train mode
96        Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

D:\anaconda\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:425: The 'val_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=11` in the `DataLoader` to improve performance.
D:\anaconda\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:425: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=11` in the `DataLoader` to improve performance.


Training: |          | 0/? [00:00<?, ?it/s]

RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn

In [52]:
train_acc = trainer.test(lightning_model, dataloaders=train_loader, ckpt_path="best", verbose=False)
val_acc = trainer.test(lightning_model, dataloaders=val_loader, ckpt_path="best", verbose=False)
test_acc = trainer.test(lightning_model, dataloaders=test_loader, ckpt_path="best", verbose=False)

Restoring states from the checkpoint path at logs/my-model\version_1\checkpoints\epoch=0-step=2917.ckpt


RuntimeError: Error(s) in loading state_dict for CustomLightningModule:
	Missing key(s) in state_dict: "model.distilbert.transformer.layer.0.attention.q_lin.weight", "model.distilbert.transformer.layer.0.attention.q_lin.bias", "model.distilbert.transformer.layer.0.attention.v_lin.weight", "model.distilbert.transformer.layer.0.attention.v_lin.bias", "model.distilbert.transformer.layer.1.attention.q_lin.weight", "model.distilbert.transformer.layer.1.attention.q_lin.bias", "model.distilbert.transformer.layer.1.attention.v_lin.weight", "model.distilbert.transformer.layer.1.attention.v_lin.bias", "model.distilbert.transformer.layer.2.attention.q_lin.weight", "model.distilbert.transformer.layer.2.attention.q_lin.bias", "model.distilbert.transformer.layer.2.attention.v_lin.weight", "model.distilbert.transformer.layer.2.attention.v_lin.bias", "model.distilbert.transformer.layer.3.attention.q_lin.weight", "model.distilbert.transformer.layer.3.attention.q_lin.bias", "model.distilbert.transformer.layer.3.attention.v_lin.weight", "model.distilbert.transformer.layer.3.attention.v_lin.bias", "model.distilbert.transformer.layer.4.attention.q_lin.weight", "model.distilbert.transformer.layer.4.attention.q_lin.bias", "model.distilbert.transformer.layer.4.attention.v_lin.weight", "model.distilbert.transformer.layer.4.attention.v_lin.bias", "model.distilbert.transformer.layer.5.attention.q_lin.weight", "model.distilbert.transformer.layer.5.attention.q_lin.bias", "model.distilbert.transformer.layer.5.attention.v_lin.weight", "model.distilbert.transformer.layer.5.attention.v_lin.bias". 
	Unexpected key(s) in state_dict: "model.distilbert.transformer.layer.0.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.0.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.0.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.0.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.0.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.0.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.0.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.0.attention.v_lin.lora.W_b", "model.distilbert.transformer.layer.1.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.1.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.1.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.1.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.1.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.1.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.1.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.1.attention.v_lin.lora.W_b", "model.distilbert.transformer.layer.2.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.2.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.2.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.2.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.2.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.2.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.2.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.2.attention.v_lin.lora.W_b", "model.distilbert.transformer.layer.3.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.3.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.3.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.3.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.3.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.3.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.3.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.3.attention.v_lin.lora.W_b", "model.distilbert.transformer.layer.4.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.4.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.4.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.4.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.4.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.4.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.4.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.4.attention.v_lin.lora.W_b", "model.distilbert.transformer.layer.5.attention.q_lin.linear.weight", "model.distilbert.transformer.layer.5.attention.q_lin.linear.bias", "model.distilbert.transformer.layer.5.attention.q_lin.lora.W_a", "model.distilbert.transformer.layer.5.attention.q_lin.lora.W_b", "model.distilbert.transformer.layer.5.attention.v_lin.linear.weight", "model.distilbert.transformer.layer.5.attention.v_lin.linear.bias", "model.distilbert.transformer.layer.5.attention.v_lin.lora.W_a", "model.distilbert.transformer.layer.5.attention.v_lin.lora.W_b". 

In [44]:
print(f"Train acc: {train_acc[0]['accuracy']*100:2.2f}%")
print(f"Val acc:   {val_acc[0]['accuracy']*100:2.2f}%")
print(f"Test acc:  {test_acc[0]['accuracy']*100:2.2f}%")

Train acc: 91.99%
Val acc:   90.08%
Test acc:  89.09%


In [ ]:
import shutil

# Cleanup checkpoint files as we don't need them later
log_dir = f"logs/my-model"
if os.path.exists(log_dir):
    shutil.rmtree(log_dir)

Time elapsed 22.09 min
Train acc: 91.99%
Val acc:   90.08%
Test acc:  89.09%
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name     | Type                                | Params | Mode
-------------------------------------------------------------------------
0 | model    | DistilBertForSequenceClassification | 67.1 M | eval
1 | val_acc  | MulticlassAccuracy                  | 0      | train
2 | test_acc | MulticlassAccuracy                  | 0      | train
-------------------------------------------------------------------------
147 K     Trainable params
67.0 M    Non-trainable params
67.1 M    Total params
268.410   Total estimated model params size (MB)
26        Modules in train mode
96        Modules in eval mode

In [58]:
train_acc = trainer.test(lightning_model, dataloaders=train_loader, verbose=False)
val_acc = trainer.test(lightning_model, dataloaders=val_loader, verbose=False)
test_acc = trainer.test(lightning_model, dataloaders=test_loader, verbose=False)

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
D:\anaconda\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:476: Your `test_dataloader`'s sampler has shuffling enabled, it is strongly recommended that you turn shuffling off for val/test dataloaders.
D:\anaconda\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:425: The 'test_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=11` in the `DataLoader` to improve performance.


Testing: |          | 0/? [00:00<?, ?it/s]

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

In [59]:
print(f"Train acc: {train_acc[0]['accuracy']*100:2.2f}%")
print(f"Val acc:   {val_acc[0]['accuracy']*100:2.2f}%")
print(f"Test acc:  {test_acc[0]['accuracy']*100:2.2f}%")

Train acc: 50.13%
Val acc:   49.10%
Test acc:  49.95%
